prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1408 @@
+
+The field `msg_id` is provided as message ID. It is used to implement ordered atomic operations, which is supported by libfabric EFA provider,
+and is required by some application such as MPICH.
+
+The workflows of emulated fetch/compare atomic are the same, as illustrated in the following diagram:
+
+![atomic_fetch_compare](atomic_fetch_compare.png)
+
+Comparing to write atomic, the differences are:
+
+first, an FETCH_RTA/COMPARE_RTA is used to initiate the communication.
+second, that responder will send an ATOMRSP (atomic response) packet back.
+
+The binary format of FETCH_RTA and COMPARE_RTA are the same.
+Table 3.14 shows the format of mandatory header of a FETCH_RTA/COMPARE_RTA packet:
+
+Table: 3.14 the format of an FETCH_RTA/COMPARE_RTA packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `msg_id`         | 4 | integer | `uint32_t` | message ID |
+| `rma_iov_count`  | 4 | integer | `uint32_t` | number of RMA iov structure |
+| `atomic_datatype`| 4 | integer | `uint32_t` | atomic data type |
+| `atomic_op`      | 4 | integer | `uint32_t` | atomic operation ID |
+| `recv_id`        | 4 | integer | `uint32_t` | ID of the receive operation on the requester side |
+| `rma_iov`        | `rma_iov_count` * 24 | array of `efa_rma_iov` | `efa_rma_iov[]` | remote buffer information |
+
+The differences between a FETCH_RTA and a COMPARE_RTA are:
+
+First, The value of `atomic_op` is different between FETCH_RTA and COMPARE_RTA.
+
+Second, the application data part of a COMPARE_RTA packet contains two segments of data: `buf` and `compare`.
+(see [fi_atomic](https://ofiwg.github.io/libfabric/v1.4.0/man/fi_atomic.3.html))
+
+The difference between an WRITE_RTA and an FETCH_RTA/COMPARE_RTA is that the field `pad` was replaced by `recv_id`.
+Because we are using send/receive to emulate a fetch/compare atomic operation. The requester is going to receive
+data from the responder, the field `recv_id` is the ID of the receive operation the requester side, which is to
+be included in the header of ATOMRSP packet.
+
+An ATOMRSP packet is consisted of two parts: header and application data. Table 3.15 shows the format of the header of an ATOMRSP packet:
+
+Table: 3.15 the binary format of an ATOMRSP packet header.
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `connid`         | 4 | integer | `uint32_t` | sender connection ID, optionally set |
```

Comment:
It seems odd to have connid in the middle of the packet since the REQ packets all have it as an optional item at the end

Reply:
This is because the space used to be a padding space, it is re-purposed to store connid. Clarified in new revision.